### PR TITLE
introduced ZooKeeperSession

### PR DESCRIPTION
**Pull request by ymatsuda**
- preventing a series of zk api calls from being issues to different ZooKeeper instances (when it is replaced by ZooKeeperClient behind the scene). This is very unlikely scenario, but still possible.
- replaced latches with a scala promise

@eishay @stkem

### CODE REVIEW

**Comment by stkem** [comment]
Nice!

File: kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala [hunk 1]
```diff
@@ -112,27 +107,9 @@ class ServiceDiscoveryImpl(
     myCluster.instanceForNode(instance.node).isDefined
   }
 
-  private def keepAlive() : Unit = if (!disableRegistration) {
```
Comment (eishay) [inline kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala:115]: id keepAlive not needed anymore?

**Comment by eishay** [comment]
nice, like the sessions

File: kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala (hunk 1 above)
Comment (ymatsuda) [inline kifi-backend/modules/common/app/com/keepit/common/zookeeper/ServiceDiscovery.scala:115]: I forgot to mention about it. I don't need it in theory. We have onConnected handlers to take care of registration. When new ZooKeeper instance is created and connected to ZK servers, we call onConnected handlers.

**Pull request merged**